optimum/habana/transformers/fastddp.py

```diff
@@ -0,0 +1,119 @@
+# coding=utf-8
+# Copyright 2023 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+    Fast and lightweight alternative of DistributeDataParallel for Habana Gaudi
```

Comment:
Is the goal of this implementation to completely replace `torch.nn.parallel.DistributedDataParallel`? Or is it just for using HPU graphs?

Reply:
It's a drop-in replacement. FastDDP uses HPU graphs internally (for packing/unpacking gradients to/from the fusion buffer), but it is independent from the fact, that FW+BWD is recorded using HPU graph.

Basically, DDP usage (with a default bucket of 230 MB of fp32) incurs a cost of roughly 14 ms per step. FastDDP works like DDP with a bucket of 640 MB with a bf16 hook, with a static set of gradients) and costs about 7 ms. It is also possible to use AutoGrad hooks to automatically pack the gradients into the fusion buffer directly from the backward pass, but it would interfere with the existing logic too much (especially with multiple accumulations per optimizer step).

I assumed that the usage of HPU graphs in FastDDP is unconditional, as there are 395 gradients and twice as much additional memory copies. I decided to replace DDP with FastDDP to minimalize all the uninteresting switches and parameters passed to the script.

Yet another alternative is manually invoking all_reduce(fp32) on gradients separately, but it yields similar perf to DDP.

Reply:
Okay sounds good!
So if I understand correctly, the only use case where FastDDP could fail and not DDP is when the set of gradients evolves during training right?